server/src/test/java/org/opensearch/indices/replication/checkpoint/PublishCheckpointActionTests.java

```diff
@@ -0,0 +1,159 @@
+        final IndexShard indexShard = mock(IndexShard.class);
+        when(indexService.getShard(id)).thenReturn(indexShard);
+
+        final ShardId shardId = new ShardId(index, id);
+        when(indexShard.shardId()).thenReturn(shardId);
+
+        final RecoverySettings recoverySettings = new RecoverySettings(Settings.EMPTY, clusterService.getClusterSettings());
+
+        final PublishCheckpointAction action = new PublishCheckpointAction(
+            Settings.EMPTY,
+            transportService,
+            clusterService,
+            indicesService,
+            threadPool,
+            shardStateAction,
+            new ActionFilters(Collections.emptySet())
+        );
+
+        final ReplicationCheckpoint checkpoint = new ReplicationCheckpoint(indexShard.shardId(), 1111, 111, 11, 1);
+
+        final PublishCheckpointRequest request = new PublishCheckpointRequest(checkpoint);
+
+        final CountDownLatch latch = new CountDownLatch(1);
+        action.shardOperationOnPrimary(request, indexShard, new LatchedActionListener<>(ActionTestUtils.assertNoFailureListener(result -> {
+            // we should forward the request containing the current retention leases to the replica
+            assertThat(result.replicaRequest(), sameInstance(request));
+        }), latch));
+        latch.await();
+
+    }
+
+    public void testPublishCheckpointActionOnReplica() {
+        final IndicesService indicesService = mock(IndicesService.class);
+
+        final Index index = new Index("index", "uuid");
+        final IndexService indexService = mock(IndexService.class);
+        when(indicesService.indexServiceSafe(index)).thenReturn(indexService);
+
+        final int id = randomIntBetween(0, 4);
+        final IndexShard indexShard = mock(IndexShard.class);
+        when(indexService.getShard(id)).thenReturn(indexShard);
+
+        final ShardId shardId = new ShardId(index, id);
+        when(indexShard.shardId()).thenReturn(shardId);
+
+        final RecoverySettings recoverySettings = new RecoverySettings(Settings.EMPTY, clusterService.getClusterSettings());
+
+
+        final PublishCheckpointAction action = new PublishCheckpointAction(
+            Settings.EMPTY,
+            transportService,
+            clusterService,
+            indicesService,
+            threadPool,
+            shardStateAction,
+            new ActionFilters(Collections.emptySet())
+        );
+
+        final ReplicationCheckpoint checkpoint = new ReplicationCheckpoint(indexShard.shardId(), 1111, 111, 11, 1);
+
+        final PublishCheckpointRequest request = new PublishCheckpointRequest(checkpoint);
+
+        final PlainActionFuture<TransportReplicationAction.ReplicaResult> listener = PlainActionFuture.newFuture();
+        action.shardOperationOnReplica(request, indexShard, listener);
+        final TransportReplicationAction.ReplicaResult result = listener.actionGet();
+
+        // the result should indicate success
```

Comment:
Will this listener ever not complete with success?

Reply:
the failure is not handled in this method. Had chat with @mch2 regarding this.